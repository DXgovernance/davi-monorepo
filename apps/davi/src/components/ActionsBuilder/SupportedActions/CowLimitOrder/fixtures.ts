import { DecodedCall, SupportedAction } from 'components/ActionsBuilder/types';
import { BigNumber, utils } from 'ethers';
import { MOCK_ADDRESS } from 'hooks/Guilds/ens/fixtures';
import CowSettlement from 'contracts/CowSettlement.json';

const CowSettlementContract = new utils.Interface(CowSettlement.abi);

export const limitOrderDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.COW_SWAP_LIMIT_ORDER,
  function: CowSettlementContract.getFunction('setPreSignature'),
  to: '0xfFb1cd0F95368DDd06D556161c5D3d9f0f4Fe6d2',
  value: BigNumber.from(0),
  args: {
    orderUid: '',
    signed: '',
  },
  optionalProps: {
    sellAmount: BigNumber.from('5000000000000000000'),
    buyToken: MOCK_ADDRESS,
    sellToken: MOCK_ADDRESS,
    functionSignature: '',
  },
};
