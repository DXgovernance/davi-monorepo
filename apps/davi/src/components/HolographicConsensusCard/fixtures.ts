import { BigNumber } from 'ethers';
import { IStakes } from './types';

export const mockProposalStakeDetails: IStakes = {
  for: [
    {
      id: '1',
      staker: '0xaf8eb8c3a5d9d900aa0b98e3df0bcc17d3c5f698',
      amount: BigNumber.from('100000000000000000000'),
    },
    {
      id: '2',
      staker: '0xc5b20ade9c9cd5e0cc087c62b26b815a4bc1881f',
      amount: BigNumber.from('100000000000000000000'),
    },
    {
      id: '3',
      staker: '0xc5b20ade9c9cd5e0cc087c62b26b815a4bc1881f',
      amount: BigNumber.from('336000000000000000000'),
    },
    {
      id: '4',
      staker: '0xc5b20ade9c9cd5e0cc087c62b26b815a4bc1881f',
      amount: BigNumber.from('236600000000000000000'),
    },
    {
      id: '5',
      staker: '0xc5b20ade9c9cd5e0cc087c62b26b815a4bc1881f',
      amount: BigNumber.from('47138000000000000000'),
    },
  ],
  against: [
    {
      id: '6',
      staker: '0x84eeb305da0a4309a696d43de9f79f04e66eb4f8',
      amount: BigNumber.from('100000000000000000000'),
    },
  ],
};

export const mockProposalStakeDetailsEmpty = {
  for: [],
  against: [],
};
