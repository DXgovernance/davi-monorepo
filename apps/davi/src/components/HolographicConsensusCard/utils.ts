import { BigNumber, FixedNumber } from 'ethers';
import { Proposal } from 'types/types.guilds.d';
import { IStakes } from './types';

export const calculateSpeedometerValue = (
  totalStaked: Proposal['totalStaked']
): number => {
  if (!totalStaked) return 0;

  const stakedAgainst = totalStaked[0];
  const stakedFor = totalStaked[1];

  const stakeDifference = stakedFor.sub(stakedAgainst);
  let percentage: number;

  if (stakeDifference.gt(0)) {
    percentage = stakeDifference.mul(100).div(stakedFor).toNumber();
  } else if (stakeDifference.eq(0)) {
    percentage = 0;
  } else {
    percentage = stakeDifference.mul(100).div(stakedAgainst).toNumber();
  }
  return 5000 + percentage * 50;
};

export const calculateBNPercentage = (
  value: BigNumber,
  percentage: string | number
) => {
  // percentage should be an integer (before being divided by 100)
  // if a percentage has decimals, it trims them
  // TODO: handle decimals?
  if (typeof percentage === 'string') {
    if (percentage.includes('.')) {
      const percentageSplit = percentage.split('.');
      percentage = percentageSplit[0];
    }
  }
  if (typeof percentage === 'number') {
    if (percentage % 1 !== 0) {
      percentage = Math.trunc(percentage);
    }
  }

  const valueFixed = FixedNumber.fromValue(value);
  const percentageFixed = FixedNumber.fromValue(BigNumber.from(percentage));
  const hundredFixed = FixedNumber.fromValue(BigNumber.from(100));

  const resultFixed = valueFixed
    .mulUnsafe(percentageFixed)
    .divUnsafe(hundredFixed);

  const resultFixedInteger = resultFixed.toString().split('.')[0];
  const resultBN = BigNumber.from(resultFixedInteger);
  return resultBN;
};

export const checkUserStakeOption = (
  userAddress: string,
  stakeDetails: IStakes
) => {
  const userStakeFor = stakeDetails.for.find(
    stake => stake.staker.toLowerCase() === userAddress.toLowerCase()
  );
  if (userStakeFor) return 'for';

  const userStakeAgainst = stakeDetails.against.find(
    stake => stake.staker.toLowerCase() === userAddress.toLowerCase()
  );
  if (userStakeAgainst) return 'against';

  return null;
};

export const calculatePotentialReward = (
  stakes: IStakes,
  userCurrentStake: BigNumber,
  userAddress: string,
  userOption: BigNumber,
  totalStaked: Proposal['totalStaked'],
  daoBounty: BigNumber
) => {
  if (userOption === undefined || userOption === null) return BigNumber.from(0);
  const userStakedOptionIndex: number = userOption.toNumber() - 1;
  let userStakedOptionString: string;
  if (userStakedOptionIndex === 0) userStakedOptionString = 'against';
  if (userStakedOptionIndex === 1) userStakedOptionString = 'for';
  if (!userStakedOptionString)
    throw new Error('Invalid userStakedOptionString');

  const stakedInUserVote: BigNumber =
    totalStaked[userStakedOptionIndex].add(userCurrentStake);
  const stakedInUserVoteFixed: FixedNumber =
    FixedNumber.fromValue(stakedInUserVote);

  if (stakedInUserVote.isZero()) return BigNumber.from(0);

  const totalStakedWithoutDaoBounty = userCurrentStake
    .add(totalStaked[0])
    .add(totalStaked[1])
    .sub(daoBounty);

  // Calculate what the user staked
  const previouslyStakedByUser = stakes[userStakedOptionString].reduce(
    (acc: BigNumber, cur) => {
      if (cur.staker.toLowerCase() === userAddress.toLowerCase()) {
        return acc.add(cur.amount);
      }
      return acc;
    },
    BigNumber.from(0)
  );

  const totalUserStake = previouslyStakedByUser.add(userCurrentStake);

  // Calculate the coefficient for the base reward

  const initialRewardCoefficient = FixedNumber.fromValue(
    totalStakedWithoutDaoBounty
  ).divUnsafe(stakedInUserVoteFixed);

  const initialReward = FixedNumber.fromValue(totalUserStake).mulUnsafe(
    initialRewardCoefficient
  );

  const initialRewardBN = BigNumber.from(
    initialReward.toString().split('.')[0]
  );

  // If the user voted for "NO", then the reward is just the initial reward
  if (userOption.eq(1)) return initialRewardBN;

  // If the user voted for "YES", then the reward is the initial reward + a percentage of the DAO bounty
  const daoBountyRewardCoefficient = FixedNumber.fromValue(daoBounty).divUnsafe(
    stakedInUserVoteFixed
  );

  const daoBountyReward = FixedNumber.fromValue(totalUserStake).mulUnsafe(
    daoBountyRewardCoefficient
  );

  const daoBountyRewardBN = BigNumber.from(
    daoBountyReward.toString().split('.')[0]
  );

  return initialRewardBN.add(daoBountyRewardBN);
};
