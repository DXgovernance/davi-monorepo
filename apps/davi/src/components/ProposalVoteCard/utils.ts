import { VoteOnProposalProps } from './types';

export const checkVotingPower = ({
  votingPowerAtProposalSnapshot,
  votingPowerAtProposalCurrentSnapshot,
}: VoteOnProposalProps) => {
  const hasNoVotingPower =
    votingPowerAtProposalSnapshot &&
    Number(votingPowerAtProposalSnapshot?.toString()) <= 0;

  const hasVotingPowerAtCurrentSnapshot =
    votingPowerAtProposalCurrentSnapshot &&
    Number(votingPowerAtProposalCurrentSnapshot?.toString()) > 0;
  return {
    hasNoVotingPower,
    hasVotingPowerAtCurrentSnapshot,
  };
};
