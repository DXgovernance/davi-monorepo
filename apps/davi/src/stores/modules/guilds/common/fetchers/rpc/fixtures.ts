export const MOCK_CONTRACT_ADDRESS =
  '0x0000000000000000000000000000000000000001';
export const MOCK_GUILD_ADDRESS = '0x0000000000000000000000000000000000000002';
