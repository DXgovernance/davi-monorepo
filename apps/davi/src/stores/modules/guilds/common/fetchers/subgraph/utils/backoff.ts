import { useBlockNumber } from 'wagmi';

export const useBackoff = (refetch: () => any) => {
  const { data: block } = useBlockNumber({
    watch: true,
  });
  let currentBlock = 0;

  // Backoff and retry logic
  const backoff = (fun, successFun, failureFun, exponent) => {
    setTimeout(async () => {
      const { data } = await fun();
      if (currentBlock < 1) {
        console.log('Still fetching current block');
      } else if (data?._meta.block.number > currentBlock) {
        successFun();
      } else if (exponent <= 20) {
        backoff(fun, successFun, failureFun, exponent + 1);
      } else {
        failureFun();
      }
    }, Math.pow(2, exponent) + Math.random() * 10000);
  };

  return async () => {
    currentBlock = block;
    backoff(
      refetch,
      () => (currentBlock = 0),
      () => console.log('Failed to fetch new data'),
      0
    );
  };
};
