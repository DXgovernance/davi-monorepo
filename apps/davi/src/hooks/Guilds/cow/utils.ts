import { GNOSIS_ID, LOCALHOST_ID, MAINNET_ID } from 'utils';

export const isAvailableOnCow = (chainId: number) => {
  return [MAINNET_ID, GNOSIS_ID, LOCALHOST_ID].includes(chainId);
};
