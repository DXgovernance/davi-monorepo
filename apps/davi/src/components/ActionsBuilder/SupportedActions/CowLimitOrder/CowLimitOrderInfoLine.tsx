import { ActionViewProps } from '..';
import { Segment } from '../common/infoLine';
import useBigNumberToNumber from 'hooks/Guilds/conversions/useBigNumberToNumber';
import { useCallback, useEffect, useMemo, useState } from 'react';
import { FiArrowRight } from 'react-icons/fi';
import { useTokenList } from 'hooks/Guilds/tokens/useTokenList';
import { useNetwork } from 'wagmi';
import { IoSwapVerticalOutline } from 'react-icons/io5';
import { CowQuote, useCow } from 'hooks/Guilds/cow/useCow';
import { BigNumber } from 'ethers';
import { ERC20_APPROVE_SIGNATURE } from 'utils/constants';
import { useTranslation } from 'react-i18next';

const CowLimitOrderInfoLine: React.FC<ActionViewProps> = ({
  decodedCall,
}) => {
  const { t } = useTranslation();

  const { chain } = useNetwork();
  const { tokens } = useTokenList(chain?.id, false);
  const [order, setOrder] = useState<CowQuote>(null);

  const findTokenByAddress = useCallback(
    (address: string) => {
      return tokens.find(
        token => token.address?.toLowerCase() === address?.toLowerCase()
      );
    },
    [tokens]
  );

  const { getOrder } = useCow();

  const parsedData = useMemo(() => {
    if (!decodedCall) return null;

    const buyToken = tokens.find(
      token =>
        token.address?.toLowerCase() ===
        decodedCall.optionalProps?.buyToken?.address?.toLowerCase()
    );
    const sellToken = tokens.find(
      token =>
        token.address?.toLowerCase() ===
        decodedCall.optionalProps?.sellToken?.address?.toLowerCase()
    );

    return {
      orderId: decodedCall.args.orderUid,
      buyToken,
      sellToken,
      buyAmount: decodedCall.optionalProps.buyAmount,
      sellAmount: decodedCall.optionalProps.sellAmount,
      functionSignature: decodedCall.optionalProps.functionSignature,
    };
  }, [decodedCall, tokens]);

  useEffect(() => {
    if (!parsedData?.orderId) return;

    const fetchOrder = async () => {
      try {
        const order = await getOrder(parsedData?.orderId);
        setOrder(order);
      } catch (error) {
        console.log('error retrieving order', error);
      }
    };

    fetchOrder();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [parsedData?.orderId]);

  const roundedBuyBalance = useBigNumberToNumber(
    order?.buyAmount ? BigNumber.from(order?.buyAmount) : null,
    order?.buyToken ? findTokenByAddress(order?.buyToken).decimals : null,
    6
  );

  const roundedSellBalance = useBigNumberToNumber(
    order?.sellAmount ? BigNumber.from(order?.sellAmount) : null,
    order?.sellToken ? findTokenByAddress(order?.sellToken)?.decimals : null,
    6
  );

  return (
    <>
      <Segment>
        <IoSwapVerticalOutline size={16} />
      </Segment>
      <Segment>
        {!(parsedData?.functionSignature === ERC20_APPROVE_SIGNATURE)
          ? `${t('actionBuilder.config.limitOrder')} ${roundedSellBalance} ${
              parsedData?.sellToken?.symbol
            }`
          : t('actionBuilder.approval.setLimitOrderApproval')}
      </Segment>
      <Segment>
        {!(parsedData?.functionSignature === ERC20_APPROVE_SIGNATURE) && (
          <FiArrowRight />
        )}
      </Segment>
      <Segment>
        {roundedBuyBalance} {parsedData?.buyToken?.symbol}
      </Segment>
    </>
  );
};

export default CowLimitOrderInfoLine;
