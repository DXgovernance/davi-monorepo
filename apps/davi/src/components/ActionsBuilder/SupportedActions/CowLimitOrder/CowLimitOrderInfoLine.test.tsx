import { mockChain } from 'components/Web3Modals/fixtures';
import { BigNumber } from 'ethers';
import { render } from 'utils/tests';
import CowLimitOrderInfoLine from './CowLimitOrderInfoLine';
import { limitOrderDecodedCallMock } from './fixtures';

const mockBigNumber = BigNumber.from(100000000);

jest.mock('hooks/Guilds/ens/useENSAvatar', () => ({
  __esModule: true,
  default: () => ({
    avatarUri: 'test',
    imageUrl: 'test',
    ensName: 'test.eth',
  }),
}));

jest.mock('hooks/Guilds/erc20/useERC20Info', () => ({
  useERC20Info: () => ({
    name: 'Test ERC20',
    symbol: 'TEST',
    decimals: 18,
    totalSupply: mockBigNumber,
  }),
}));

jest.mock('wagmi', () => ({
  useContractRead: () => ({ data: '' }),
  useNetwork: () => ({ chain: mockChain, chains: [mockChain] }),
  useEnsResolver: () => ({
    data: {
      name: 'name.eth',
      address: '0x0000000000000000000000000000000000000000',
      contentHash: '0x0',
    },
  }),
}));

describe('CowLimitOrderInfoLine', () => {
  it('Should match snapshot', () => {
    const { container } = render(
      <CowLimitOrderInfoLine decodedCall={limitOrderDecodedCallMock} />
    );
    expect(container).toMatchSnapshot();
  });
});
