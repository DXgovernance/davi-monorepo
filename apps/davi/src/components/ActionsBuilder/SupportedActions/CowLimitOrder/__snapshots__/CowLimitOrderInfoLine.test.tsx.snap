// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CowLimitOrderInfoLine Should match snapshot 1`] = `
.c0 {
  margin-right: 0.5rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div>
   
  <span
    class="c0"
  >
    <svg
      fill="currentColor"
      height="16"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 512 512"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M464 208L352 96 240 208m112-94.87V416M48 304l112 112 112-112m-112 94V96"
        fill="none"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="32"
      />
    </svg>
  </span>
  <span
    class="c0"
  >
    actionBuilder.config.limitOrder null undefined
  </span>
  <span
    class="c0"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        x1="5"
        x2="19"
        y1="12"
        y2="12"
      />
      <polyline
        points="12 5 19 12 12 19"
      />
    </svg>
  </span>
  <span
    class="c0"
  >
     
  </span>
   
</div>
`;
